# Introduction to Random Forests

Modern machine learning can be distilled down to a couple of key techniques that are of very wide applicability. Recent studies have shown that the vast majority of datasets can be best modeled with just two methods:

- *Ensembles of decision trees* (i.e. Random Forests and Gradient Boosting Machines), mainly for structured data (such as you might find in a database table at most companies)
- *Multi-layered neural networks learnt with SGD* (i.e. shallow and/or deep learning), mainly for unstructured data (such as audio, vision, and natural language)

In this course we'll be doing a deep dive into random forests, and simple models learnt with SGD. We'll be learning about gradient boosting and deep learning in part 2.

The dataset that we will use for this part of the course is from the "Blue Book for Bulldozers" Kaggle competition, located [here](https://www.kaggle.com/c/bluebook-for-bulldozers). The predictive modelling challenge was to predict the sale price of a heavy piece of machinery at auction based on its usage, equipment type and configuration.The evaluation metric for this competition was the RMSLE (root mean squared log error) between the actual and predicted auction prices.

### Imports

In [7]:
%load_ext autoreload
%autoreload 2
%matplotlib inline

The last import line allows us to display plotted data within the notebook

In [8]:
from fastai.imports import *
from fastai.structured import *

from pandas_summary import DataFrameSummary
from sklearn.ensemble import RandomForestRegressor, RandomForestClassifier
from IPython.display import display

from sklearn import metrics

In [9]:
?display

Adding a question mark before a function will tell you what the function does

In [10]:
PATH = "data/bulldozers/"
!ls {PATH}

Train.csv  bulldozers.zip


## The data


Kaggle provides info about some of the fields of our dataset; on the [Kaggle Data info](https://www.kaggle.com/c/bluebook-for-bulldozers/data) page they say the following:

For this competition, we are predicting the sale price of bulldozers sold at auctions. The data for this competition is split into three parts:

- **Train.csv** is the training set, which contains data through the end of 2011.
- **Valid.csv** is the validation set, which contains data from January 1, 2012 - April 30, 2012. You make predictions on this set throughout the majority of the competition. Your score on this set is used to create the public leaderboard.
- **Test.csv** is the test set, which won't be released until the last week of the competition. It contains data from May 1, 2012 - November 2012. Your score on the test set determines your final rank for the competition.

The key fields are in train.csv are:

- SalesID: the unique identifier of the sale
- MachineID: the unique identifier of a machine.  A machine can be sold multiple times
- saleprice: what the machine sold for at auction (only provided in train.csv)
- saledate: the date of the sale

Clarification on new Python 3.6 syntax:

In [11]:
name = 'Snehan'
f'Hello {name}'

'Hello Snehan'

In [12]:
name = 'Snehan'
age = 24
f'Hello {name.upper()}, you are {age}'

'Hello SNEHAN, you are 24'

We shall now read the import the data. Note that 'df' stands for dataframe.

In [13]:
df_raw = pd.read_csv(f'{PATH}Train.csv', low_memory = False,
                    parse_dates = ["saledate"])

In [14]:
df_raw

SalesID  SalePrice  MachineID  ModelID  datasource  auctioneerID  \
0       1139246      66000     999089     3157         121           3.0   
1       1139248      57000     117657       77         121           3.0   
2       1139249      10000     434808     7009         121           3.0   
3       1139251      38500    1026470      332         121           3.0   
4       1139253      11000    1057373    17311         121           3.0   
...         ...        ...        ...      ...         ...           ...   
401120  6333336      10500    1840702    21439         149           1.0   
401121  6333337      11000    1830472    21439         149           1.0   
401122  6333338      11500    1887659    21439         149           1.0   
401123  6333341       9000    1903570    21435         149           2.0   
401124  6333342       7750    1926965    21435         149           2.0   

        YearMade  MachineHoursCurrentMeter UsageBand   saledate  ...  \
0           2004                      68.0       Low 2006-11-16  ...   
1           1996                    4640.0       Low 2004-03-26  ...   
2           2001                    2838.0      High 2004-02-26  ...   
3           2001                    3486.0      High 2011-05-19  ...   
4           2007                     722.0    Medium 2009-07-23  ...   
...          ...                       ...       ...        ...  ...   
401120      2005                       NaN       NaN 2011-11-02  ...   
401121      2005                       NaN       NaN 2011-11-02  ...   
401122      2005                       NaN       NaN 2011-11-02  ...   
401123      2005                       NaN       NaN 2011-10-25  ...   
401124      2005                       NaN       NaN 2011-10-25  ...   

       Undercarriage_Pad_Width         Stick_Length                Thumb  \
0                          NaN                  NaN                  NaN   
1                          NaN                  NaN                  NaN   
2                          NaN                  NaN                  NaN   
3                          NaN                  NaN                  NaN   
4                          NaN                  NaN                  NaN   
...                        ...                  ...                  ...   
401120     None or Unspecified  None or Unspecified  None or Unspecified   
401121     None or Unspecified  None or Unspecified  None or Unspecified   
401122     None or Unspecified  None or Unspecified  None or Unspecified   
401123     None or Unspecified  None or Unspecified  None or Unspecified   
401124     None or Unspecified  None or Unspecified  None or Unspecified   

            Pattern_Changer Grouser_Type Backhoe_Mounting Blade_Type  \
0                       NaN          NaN              NaN        NaN   
1                       NaN          NaN              NaN        NaN   
2                       NaN          NaN              NaN        NaN   
3                       NaN          NaN              NaN        NaN   
4                       NaN          NaN              NaN        NaN   
...                     ...          ...              ...        ...   
401120  None or Unspecified       Double              NaN        NaN   
401121  None or Unspecified       Double              NaN        NaN   
401122  None or Unspecified       Double              NaN        NaN   
401123  None or Unspecified       Double              NaN        NaN   
401124  None or Unspecified       Double              NaN        NaN   

       Travel_Controls Differential_Type Steering_Controls  
0                  NaN          Standard      Conventional  
1                  NaN          Standard      Conventional  
2                  NaN               NaN               NaN  
3                  NaN               NaN               NaN  
4                  NaN               NaN               NaN  
...                ...               ...               ...  
401120             NaN   

In [15]:
?pd.option_context

In [16]:
def display_all(df):
    with pd.option_context("display.max_rows", 1000):
        with pd.option_context("display.max_columns", 1000):
            display(df)

In [17]:
display_all(df_raw.tail().T)

401120  \
SalesID                                                             6333336   
SalePrice                                                             10500   
MachineID                                                           1840702   
ModelID                                                               21439   
datasource                                                              149   
auctioneerID                                                              1   
YearMade                                                               2005   
MachineHoursCurrentMeter                                                NaN   
UsageBand                                                               NaN   
saledate                                                2011-11-02 00:00:00   
fiModelDesc                                                           35NX2   
fiBaseModel                                                              35   
fiSecondaryDesc                                                          NX   
fiModelSeries                                                             2   
fiModelDescriptor                                                       NaN   
ProductSize                                                            Mini   
fiProductClassDesc        Hydraulic Excavator, Track - 3.0 to 4.0 Metric...   
state                                                              Maryland   
ProductGroup                                                            TEX   
ProductGroupDesc                                           Track Excavators   
Drive_System                                                            NaN   
Enclosure                                                             EROPS   
Forks                                                                   NaN   
Pad_Type                                                                NaN   
Ride_Control                                                            NaN   
Stick                                                                   NaN   
Transmission                                                            NaN   
Turbocharged                                                            NaN   
Blade_Extension                                                         NaN   
Blade_Width                                                             NaN   
Enclosure_Type                                                          NaN   
Engine_Horsepower                                                       NaN   
Hydraulics                                                        Auxiliary   
Pushblock                                                               NaN   
Ripper                                                                  NaN   
Scarifier                                                               NaN   
Tip_Control                                                             NaN   
Tire_Size                                                               NaN   
Coupler                                                 None or Unspecified   
Coupler_System                                                          NaN   
Grouser_Tracks                                                          NaN   
Hydraulics_Flow                                                         NaN   
Track_Type                                                            Steel   
Undercarriage_Pad_Width                                 None or Unspecified   
Stick_Length                                            None or Unspecified   
Thumb                                                   None or Unspecified   
Pattern_Changer                                         None or Unspecified   
Grouser_Type                                                         Double   
Backhoe_Mounting                                                        NaN   
Blade_Type                                                              NaN   
Travel_Controls                         

In [18]:
display_all(df_raw.tail().transpose())

401120  \
SalesID                                                             6333336   
SalePrice                                                             10500   
MachineID                                                           1840702   
ModelID                                                               21439   
datasource                                                              149   
auctioneerID                                                              1   
YearMade                                                               2005   
MachineHoursCurrentMeter                                                NaN   
UsageBand                                                               NaN   
saledate                                                2011-11-02 00:00:00   
fiModelDesc                                                           35NX2   
fiBaseModel                                                              35   
fiSecondaryDesc                                                          NX   
fiModelSeries                                                             2   
fiModelDescriptor                                                       NaN   
ProductSize                                                            Mini   
fiProductClassDesc        Hydraulic Excavator, Track - 3.0 to 4.0 Metric...   
state                                                              Maryland   
ProductGroup                                                            TEX   
ProductGroupDesc                                           Track Excavators   
Drive_System                                                            NaN   
Enclosure                                                             EROPS   
Forks                                                                   NaN   
Pad_Type                                                                NaN   
Ride_Control                                                            NaN   
Stick                                                                   NaN   
Transmission                                                            NaN   
Turbocharged                                                            NaN   
Blade_Extension                                                         NaN   
Blade_Width                                                             NaN   
Enclosure_Type                                                          NaN   
Engine_Horsepower                                                       NaN   
Hydraulics                                                        Auxiliary   
Pushblock                                                               NaN   
Ripper                                                                  NaN   
Scarifier                                                               NaN   
Tip_Control                                                             NaN   
Tire_Size                                                               NaN   
Coupler                                                 None or Unspecified   
Coupler_System                                                          NaN   
Grouser_Tracks                                                          NaN   
Hydraulics_Flow                                                         NaN   
Track_Type                                                            Steel   
Undercarriage_Pad_Width                                 None or Unspecified   
Stick_Length                                            None or Unspecified   
Thumb                                                   None or Unspecified   
Pattern_Changer                                         None or Unspecified   
Grouser_Type                                                         Double   
Backhoe_Mounting                                                        NaN   
Blade_Type                                                              NaN   
Travel_Controls                         

In [19]:
display_all(df_raw.describe(include='all').T)

count unique  \
SalesID                   401125    NaN   
SalePrice                 401125    NaN   
MachineID                 401125    NaN   
ModelID                   401125    NaN   
datasource                401125    NaN   
auctioneerID              380989    NaN   
YearMade                  401125    NaN   
MachineHoursCurrentMeter  142765    NaN   
UsageBand                  69639      3   
saledate                  401125   3919   
fiModelDesc               401125   4999   
fiBaseModel               401125   1950   
fiSecondaryDesc           263934    175   
fiModelSeries              56908    122   
fiModelDescriptor          71919    139   
ProductSize               190350      6   
fiProductClassDesc        401125     74   
state                     401125     53   
ProductGroup              401125      6   
ProductGroupDesc          401125      6   
Drive_System              104361      4   
Enclosure                 400800      6   
Forks                     192077      2   
Pad_Type                   79134      4   
Ride_Control              148606      3   
Stick                      79134      2   
Transmission              183230      8   
Turbocharged               79134      2   
Blade_Extension            25219      2   
Blade_Width                25219      6   
Enclosure_Type             25219      3   
Engine_Horsepower          25219      2   
Hydraulics                320570     12   
Pushblock                  25219      2   
Ripper                    104137      4   
Scarifier                  25230      2   
Tip_Control                25219      3   
Tire_Size                  94718     17   
Coupler                   213952      3   
Coupler_System             43458      2   
Grouser_Tracks             43362      2   
Hydraulics_Flow            43362      3   
Track_Type                 99153      2   
Undercarriage_Pad_Width    99872     19   
Stick_Length               99218     29   
Thumb                      99288      3   
Pattern_Changer            99218      3   
Grouser_Type               99153      3   
Backhoe_Mounting           78672      2   
Blade_Type                 79833     10   
Travel_Controls            79834      7   
Differential_Type          69411      4   
Steering_Controls          69369      5   

                                                                        top  \
SalesID                                                                 NaN   
SalePrice                                                               NaN   
MachineID                                                               NaN   
ModelID                                                                 NaN   
datasource                                                              NaN   
auctioneerID                                                            NaN   
YearMade                                                                NaN   
MachineHoursCurrentMeter                                                NaN   
UsageBand                                                            Medium   
saledate                                                2009-02-16 00:00:00   
fiModelDesc                                                            310G   
fiBaseModel                                                             580   
fiSecondaryDesc                                                           C   
fiModelSeries                                                            II   
fiModelDescriptor                                                         L   
ProductSize                                                          Medium   
fiProductClassDesc        Backhoe Loader - 14.0 to 15.0 Ft Standard Digg...   
state                                                               Florida   
ProductGroup                                                            TEX   
ProductGroupDesc                                           Track Excavators   
Drive_System                                 

To avoid overfitting, many academic sources suggest that *Exploratory Data Analysis* should not be undertaken first; however, practical sources suggest doing so. Jeremy suggests the answer lies somewhere in the middle: ML-driven EDA.

As the evaluation metric for this particular project requires the RMSLE between the actual and predicted auction prices, we now take the log of the prices.

In [20]:
df_raw.SalePrice = np.log(df_raw.SalePrice)

In [21]:
df_raw.SalePrice

0         11.097410
1         10.950807
2          9.210340
3         10.558414
4          9.305651
            ...    
401120     9.259131
401121     9.305651
401122     9.350102
401123     9.104980
401124     8.955448
Name: SalePrice, Length: 401125, dtype: float64

### Creating a Random Forest

There is a difference between methods that predict continuous variables and those that predict categorical/discrete variables. In the case of the former, the ML models are called *Regression* models and in the latter, they are called *Classification* models. 

In [22]:
RandomForestRegressor

sklearn.ensemble.forest.RandomForestRegressor

In [23]:
?RandomForestRegressor

Below, we create an instance of the estimator. We then call `.fit` and pass in our independent variables - the data we want to use for prediction, followed by the dependant vairable. The `.drop` method returns a new dataframe with a list of columns or rows removed; *axis = 1* specifies the removal of a column. Note that `?` returns the documentation where `??` returns the source code.

In [24]:
m = RandomForestRegressor(n_jobs = -1) # -1 allows the estimator to use all the processors
m.fit (df_raw.drop('SalePrice', axis = 1), df_raw.SalePrice)

/home/snehansighat/anaconda3/envs/fastai/lib/python3.6/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


ValueError: could not convert string to float: 'Low'

The reason behind the error above is that the dataset contains a mix of continuous and categorical variables. When extracting our data, we told the program to parse the information in the 'saledate' column as dates. The datetype therefore is `datetime64` (64 bit) which is not a number! Here we come across our first example of feature engineering.

The following method extracts particular date fields from a complete datetime for the purpose of constructing categoricals.  One should always consider this feature extraction step when working with date-time. Without expanding the date-time into these additional fields, you can't capture any trend/cyclical behavior as a function of time at any of these granularities, e.g public holidays, events, etc.

In [25]:
df_raw.saledate

0        2006-11-16
1        2004-03-26
2        2004-02-26
3        2011-05-19
4        2009-07-23
            ...    
401120   2011-11-02
401121   2011-11-02
401122   2011-11-02
401123   2011-10-25
401124   2011-10-25
Name: saledate, Length: 401125, dtype: datetime64[ns]

In [26]:
??add_datepart

`add_datepart` converts a column from a dataframe, from datetime64 to many columns containing the information from the date. One has to access attributes such as year from the datetime method using `.dt.attribute`, we can then choose as many attributes as we feel relevant and seperating them out for use.

In [27]:
fld = df_raw.saledate

In [28]:
fld.dt.year

0         2006
1         2004
2         2004
3         2011
4         2009
          ... 
401120    2011
401121    2011
401122    2011
401123    2011
401124    2011
Name: saledate, Length: 401125, dtype: int64

In [29]:
fld.dt.dayofyear

0         320
1          86
2          57
3         139
4         204
         ... 
401120    306
401121    306
401122    306
401123    298
401124    298
Name: saledate, Length: 401125, dtype: int64

In [30]:
add_datepart(df_raw, 'saledate')
df_raw.saleYear.head()

0    2006
1    2004
2    2004
3    2011
4    2009
Name: saleYear, dtype: int64

In [31]:
df_raw.columns

Index(['SalesID', 'SalePrice', 'MachineID', 'ModelID', 'datasource',
       'auctioneerID', 'YearMade', 'MachineHoursCurrentMeter', 'UsageBand',
       'fiModelDesc', 'fiBaseModel', 'fiSecondaryDesc', 'fiModelSeries',
       'fiModelDescriptor', 'ProductSize', 'fiProductClassDesc', 'state',
       'ProductGroup', 'ProductGroupDesc', 'Drive_System', 'Enclosure',
       'Forks', 'Pad_Type', 'Ride_Control', 'Stick', 'Transmission',
       'Turbocharged', 'Blade_Extension', 'Blade_Width', 'Enclosure_Type',
       'Engine_Horsepower', 'Hydraulics', 'Pushblock', 'Ripper', 'Scarifier',
       'Tip_Control', 'Tire_Size', 'Coupler', 'Coupler_System',
       'Grouser_Tracks', 'Hydraulics_Flow', 'Track_Type',
       'Undercarriage_Pad_Width', 'Stick_Length', 'Thumb', 'Pattern_Changer',
       'Grouser_Type', 'Backhoe_Mounting', 'Blade_Type', 'Travel_Controls',
       'Differential_Type', 'Steering_Controls', 'saleYear', 'saleMonth',
       'saleWeek', 'saleDay', 'saleDayofweek', 'saleDayofyear',


As can be seen above, `Saledate` has been removed and replaced by a collection of columns such as `saleYear` and `saleWeek`. We also see below that the *UsageBand* column for example has strings stored as datatypes which turns out to be inefficient and does not provide the numeric coding required for a random forest. We therefore call `train_cats` to convert strings to pandas categories, which works as pandas will continue displaying the text categories, while treating them as numerical data internally; this can be seen using:

In [33]:
df_raw.head()

SalesID  SalePrice  MachineID  ModelID  datasource  auctioneerID  YearMade  \
0  1139246  11.097410     999089     3157         121           3.0      2004   
1  1139248  10.950807     117657       77         121           3.0      1996   
2  1139249   9.210340     434808     7009         121           3.0      2001   
3  1139251  10.558414    1026470      332         121           3.0      2001   
4  1139253   9.305651    1057373    17311         121           3.0      2007   

   MachineHoursCurrentMeter UsageBand fiModelDesc  ... saleDay saleDayofweek  \
0                      68.0       Low        521D  ...      16             3   
1                    4640.0       Low      950FII  ...      26             4   
2                    2838.0      High         226  ...      26             3   
3                    3486.0      High    PC120-6E  ...      19             3   
4                     722.0    Medium        S175  ...      23             3   

  saleDayofyear saleIs_month_end saleIs_month_start saleIs_quarter_end  \
0           320            False              False              False   
1            86            False              False              False   
2            57            False              False              False   
3           139            False              False              False   
4           204            False              False              False   

  saleIs_quarter_start saleIs_year_end saleIs_year_start saleElapsed  
0                False           False             False  1163635200  
1                False           False             False  1080259200  
2                False           False             False  1077753600  
3                False           False             False  1305763200  
4                False           False             False  1248307200  

[5 rows x 65 columns]

In [34]:
??train_cats

In [35]:
train_cats(df_raw)

In [37]:
df_raw.UsageBand.cat.categories

Index(['High', 'Low', 'Medium'], dtype='object')

In [38]:
df_raw.UsageBand.cat.set_categories(['High', 'Medium', 'Low'], ordered = True, inplace = True)

In [39]:
df_raw.UsageBand.cat.codes

0         2
1         2
2         0
3         0
4         1
         ..
401120   -1
401121   -1
401122   -1
401123   -1
401124   -1
Length: 401125, dtype: int8

In [40]:
display_all(df_raw.isnull().sum().sort_index()/len(df_raw))

Backhoe_Mounting            0.803872
Blade_Extension             0.937129
Blade_Type                  0.800977
Blade_Width                 0.937129
Coupler                     0.466620
Coupler_System              0.891660
Differential_Type           0.826959
Drive_System                0.739829
Enclosure                   0.000810
Enclosure_Type              0.937129
Engine_Horsepower           0.937129
Forks                       0.521154
Grouser_Tracks              0.891899
Grouser_Type                0.752813
Hydraulics                  0.200823
Hydraulics_Flow             0.891899
MachineHoursCurrentMeter    0.644089
MachineID                   0.000000
ModelID                     0.000000
Pad_Type                    0.802720
Pattern_Changer             0.752651
ProductGroup                0.000000
ProductGroupDesc            0.000000
ProductSize                 0.525460
Pushblock                   0.937129
Ride_Control                0.629527
Ripper                      0.740388
S

The above will add up the number of empty values for each series which we sort by their index and divide by the size of the dataset. We end up with essentially the percentage of missing values for each series.

In [41]:
os.makedirs('tmp', exist_ok = True)
df_raw.to_feather('tmp/bulldozers-raw')

Here we save this file for now so it can we be stored and accessed efficiently and then read from this last format:

In [42]:
df_raw = pd.read_feather('tmp/bulldozers-raw')

So far we have replaced categories with their numeric codes, handled missing continuous values, and split the dependent variable into a separate variable.

In [43]:
??proc_df

In [44]:
df, y, nas = proc_df(df_raw, 'SalePrice')

In [45]:
df.columns

Index(['SalesID', 'MachineID', 'ModelID', 'datasource', 'auctioneerID',
       'YearMade', 'MachineHoursCurrentMeter', 'UsageBand', 'fiModelDesc',
       'fiBaseModel', 'fiSecondaryDesc', 'fiModelSeries', 'fiModelDescriptor',
       'ProductSize', 'fiProductClassDesc', 'state', 'ProductGroup',
       'ProductGroupDesc', 'Drive_System', 'Enclosure', 'Forks', 'Pad_Type',
       'Ride_Control', 'Stick', 'Transmission', 'Turbocharged',
       'Blade_Extension', 'Blade_Width', 'Enclosure_Type', 'Engine_Horsepower',
       'Hydraulics', 'Pushblock', 'Ripper', 'Scarifier', 'Tip_Control',
       'Tire_Size', 'Coupler', 'Coupler_System', 'Grouser_Tracks',
       'Hydraulics_Flow', 'Track_Type', 'Undercarriage_Pad_Width',
       'Stick_Length', 'Thumb', 'Pattern_Changer', 'Grouser_Type',
       'Backhoe_Mounting', 'Blade_Type', 'Travel_Controls',
       'Differential_Type', 'Steering_Controls', 'saleYear', 'saleMonth',
       'saleWeek', 'saleDay', 'saleDayofweek', 'saleDayofyear',
       'saleI

In [46]:
df.head()

SalesID  MachineID  ModelID  datasource  auctioneerID  YearMade  \
0  1139246     999089     3157         121           3.0      2004   
1  1139248     117657       77         121           3.0      1996   
2  1139249     434808     7009         121           3.0      2001   
3  1139251    1026470      332         121           3.0      2001   
4  1139253    1057373    17311         121           3.0      2007   

   MachineHoursCurrentMeter  UsageBand  fiModelDesc  fiBaseModel  ...  \
0                      68.0          3          950          296  ...   
1                    4640.0          3         1725          527  ...   
2                    2838.0          1          331          110  ...   
3                    3486.0          1         3674         1375  ...   
4                     722.0          2         4208         1529  ...   

   saleDayofyear  saleIs_month_end  saleIs_month_start  saleIs_quarter_end  \
0            320             False               False               False   
1             86             False               False               False   
2             57             False               False               False   
3            139             False               False               False   
4            204             False               False               False   

   saleIs_quarter_start  saleIs_year_end  saleIs_year_start  saleElapsed  \
0                 False            False              False   1163635200   
1                 False            False              False   1080259200   
2                 False            False              False   1077753600   
3                 False            False              False   1305763200   
4                 False            False              False   1248307200   

   auctioneerID_na  MachineHoursCurrentMeter_na  
0            False                        False  
1            False                        False  
2            False                        False  
3            False                        False  
4            False                        False  

[5 rows x 66 columns]

Now that our dataset is purely numerical, we can pass it on to a random forest!

In [47]:
m = RandomForestRegressor(n_jobs = -1)
m.fit(df, y)
m.score(df, y)

/home/snehansighat/anaconda3/envs/fastai/lib/python3.6/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


0.9831148104079886

This is our value for the coefficient of determination *R²*, which is essentially the ratio between how good our model is (RMSE) compared to how good the naïve mean model is (RMSE).

Using all our data can lead to __overfitting__ as we have obtained the best accuracy on the training set; this accuracy may not carry over to new data, i.e the test set. A validation set helps diagnose this problem, as we require an independant dataset that was not used to create the model in the first place, which we create below while bearing in mind that we are attempting to somewhat replicate the test set.

In [49]:
def split_vals(a, n): return a[:n].copy(), a[n:].copy() 

# This function returns the first n rows (for the training set) 
# and the last n rows (for the validation set) 

n_valid = 12000 # Same size as Kaggle's test set
n_trn = len(df) - n_valid
raw_train, raw_valid = split_vals(df_raw, n_trn) 
df_raw.shape, raw_train.shape, raw_valid.shape

((401125, 65), (389125, 65), (12000, 65))

As we can see, our value for *R²* is 0.888 on the validation set, which implies that we are heavily overfitting on the training set, for which the *R²* is 0.983; however the relatively low RMSE of 0.25 tells us that the degree of overfitting is not too bad! 